components/eamxx/src/share/field/field_utils_impl_dot.hpp

```diff
@@ -0,0 +1,387 @@
+#ifndef SCREAM_FIELD_UTILS_IMPL_DOT_HPP
+#define SCREAM_FIELD_UTILS_IMPL_DOT_HPP
+
+#include "ekat/kokkos/ekat_kokkos_utils.hpp"
+#include "ekat/mpi/ekat_comm.hpp"
+#include "share/field/field.hpp"
+
+namespace scream {
+
+// dot product of rank 1 field with a rank N field, return a rank N-1 field
+template <typename ST>
+Field do_dot_along_rank1_dim(const int &pd, const Field &f1, const Field &f2,
+                             const ekat::Comm *co) {
+  using KT          = ekat::KokkosTypes<DefaultDevice>;
+  using RangePolicy = Kokkos::RangePolicy<Field::device_t::execution_space>;
+  using TeamPolicy  = Kokkos::TeamPolicy<Field::device_t::execution_space>;
+  using TeamMember  = typename TeamPolicy::member_type;
+  using ESU         = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+
+  const auto &l1 = f1.get_header().get_identifier().get_layout();
+
+  const auto &n2 = f2.get_header().get_identifier().name();
+  const auto &l2 = f2.get_header().get_identifier().get_layout();
+  const auto &u2 = f2.get_header().get_identifier().get_units();
+  const auto &g2 = f2.get_header().get_identifier().get_grid_name();
+
+  EKAT_REQUIRE_MSG(pd <= 5 && pd >= 0,
+                   "Error! First argument pd must be between 0 and 5.\n"
+                   "The input pd is "
+                       << pd << ", which is not accepted.\n");
+  EKAT_REQUIRE_MSG(l1.rank() == 1,
+                   "Error! Second argument f1 must be rank-1.\n"
+                   "The input f1 rank is "
+                       << l1.rank() << ", which is not accepted.\n");
+  EKAT_REQUIRE_MSG(l2.rank() <= 6,
+                   "Error! Third argument f2 must be at most rank-6.\n"
+                   "The input f2 rank is "
+                       << l2.rank() << ", which is not accepted.\n");
+  EKAT_REQUIRE_MSG(
+      l1.dim(0) == l2.dim(pd),
+      "Error! The two input fields must have the same dimension along "
+      "which we are taking the dot product.\n"
+      "The first field f1 has dimension "
+          << l1.dim(0)
+          << " while "
+             "the second field f2 has dimension "
+          << l2.dim(pd)
+          << " \n"
+             "along the provided dimension pd "
+          << pd << " .\n");
```

Comment:
Maybe add the checks above to the `dot_along_rank1_dim()` function like with the other utils.